Add option to reveal unused expectations files

xymaxim opened the pull request:
This adds the `--pm-reveal-unused-files` option to run tests in the collect-only mode and print out the unused expectations files. I find this functionality quite useful to keep a base directory clean after renaming or removing tests. Does it deserve a new plugin option? 

The standard `TerminalReporter` is replaced with a custom reporter to print only revealed unused files separated with newlines—without any headers, summary, etc. The output can be piped to commands like `mv` or `rm`.  Some plugins write reports to files while keeping the default output, but I think in our case it’ll be redundant.

Only files with `.out` and `.err` extensions are checked. This allows to ignore files like README, backup files, etc.

A small question: are the words ‘patterns’ and ‘expectations’ interchangeable? Or ‘patterns’ is a narrow term and refers to expectations containing regexp patterns? Should I rename the option to `--pm-reveal-unused-expectations`?

zaufi commented:
@xymaxim Thank you very much for the great improvement!

I've rebased your branch against the latest master, where I've added GH workflows to run linters and tests. Please look at the [type-checking issues found](https://github.com/zaufi/pytest-matcher/actions/runs/7598559361/job/20694657421?pr=14).

zaufi commented:
> Or ‘patterns’ is a narrow term and refers to expectations containing regexp patterns? 

Strictly speaking, it's correct. However, I'm not sure how these terms are really mentioned in code/docs nowadays ;) So, I think nowadays they are quite interchangeable ;)

> Should I rename the option to --pm-reveal-unused-expectations?

As for me, the current option name is also fine... Up to you...

zaufi commented:
Adding GH actions I was thinking that it'll be nice to have a CI check for the orphaned files... That means the process should exit w/a non-zero exit code. However, it'll ruin your original scenarios like `pytest --pm-reveal-unused-files | xargs git rm`. So, probably it should be another option for use in GH actions (or `pre-commit`) `--pm-check-unused-files` that will exit w/a non-zero exit code if any files have been found.

Thoughts?

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +342,48 @@ def pytest_addoption(parser) -> None:
+        if not session.items:
+            return
+
+        for alg in [_try_cli_option, _try_ini_option]:
+            patterns_base_dir, _ = alg(session.items[0]._request)  # NOQA: SLF001
+            if patterns_base_dir:
+                break
```
This code looks copy-n-pasted from another place. I'd be glad if you could make a reusable function out of it, please. Or I can do it later after we merge your work. %)

xymaxim replied:
What has puzzled me here is `_try_module_path()`. I’ve omitted it for now because as far as I understand, according to documentation, in the future the pattern files should reflect the layout (packages, modules, classes) of test files by creating the corresponding directories. 

What would the base directories look like in such a case? Each module directory has its own `/data/expected` directory?

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +342,48 @@ def pytest_addoption(parser) -> None:
+        config.pluginmanager.register(reporter, 'terminalreporter')
+
+
+class UnusedFilesReporter:
```
Not sure if this class should be "public" in this module… Renaming to `_UnusedFilesReporter` can give a clue to the reader that it's not a public API of the package.

Also, for code grouping purposes I'd prefer to move this class declaration above the `# BEGIN Pytest hooks` block (which contains hooks (function definitions) recognizable by `pytest`).

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +342,48 @@ def pytest_addoption(parser) -> None:
+
+        known_extensions = '.out', '.err'
+
+        all_paths: list[pathlib.Path] = []
+        for p in patterns_base_dir.rglob('*'):
+            if p.is_file() and p.suffix in known_extensions:
+                all_paths.append(p.resolve())  # NOQA: PERF401
+
+        collected_paths: list[pathlib.Path] = []
+        for item in session.items:
+            for fixture, ext in zip((expected_out, expected_err), known_extensions):
+                if fixture.__name__ in item.fixturenames:
+                    filename = _make_expected_filename(item._request, ext)  # NOQA: SLF001
+                    collected_paths.append(filename)
+
+        unused_paths = set(all_paths).difference(collected_paths)
```
The same code with comprehensions makes no need to suppress `ruff` issues :) Also, simplify set difference a little…
```suggestion
        all_paths = {
            p.resolve()
            for p in patterns_base_dir.rglob('*')
            if p.is_file() and p.suffix in known_extensions
          }

        collected_paths = {
            _make_expected_filename(item._request, ext)
            for item in session.items
            for fixture, ext in zip((expected_out, expected_err), known_extensions)
            if fixture.__name__ in item.fixturenames
          }

        unused_paths = all_paths - collected_paths
```

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +342,48 @@ def pytest_addoption(parser) -> None:
+        for item in session.items:
+            for fixture, ext in zip((expected_out, expected_err), known_extensions):
+                if fixture.__name__ in item.fixturenames:
+                    filename = _make_expected_filename(item._request, ext)  # NOQA: SLF001
```
Oh... accessing a private member is not really nice :() Would be great to find some other way if possible %()

xymaxim replied:
Yeah, unfortunately, I don't see any other way.  There's no such hook that takes an argument of `pytest.FixtureRequest` directly. I think they made the `_request` attribute private because it's not immediately available to all hooks.

xymaxim commented:
> @xymaxim Thank you very much for the great improvement!
> 
> I've rebased your branch against the latest master, where I've added GH workflows to run linters and tests. Please look at the [type-checking issues found](https://github.com/zaufi/pytest-matcher/actions/runs/7598559361/job/20694657421?pr=14).
> 
> Please note you need the following:
> 
> ```
> $ git fetch
> $ git reset --hard FETCH_HEAD
> ```

Thanks, done. Fewer worries with linters and tests now.

xymaxim commented:
> Please look at the [type-checking issues found](https://github.com/zaufi/pytest-matcher/actions/runs/7598559361/job/20694657421?pr=14).

Well, yes, the `Item` objects are actually `Function` class instances.

I’ve silenced them, as in the pytest codebase (see, e.g. [this](https://github.com/pytest-dev/pytest/blob/44bf7a2ec0b52b60089ebd8439f64487df079214/src/_pytest/runner.py#L123) file).

xymaxim commented:
> Adding GH actions I was thinking that it'll be nice to have a CI check for the orphaned files... That means the process should exit w/a non-zero exit code. However, it'll ruin your original scenarios like `pytest --pm-reveal-unused-files | xargs git rm`. So, probably it should be another option for use in GH actions (or `pre-commit`) `--pm-check-unused-files` that will exit w/ a non-zero exit code to fail a build if any files have been found.
> 
> Thoughts?

Actually, pytest doesn’t fail a pipe—it’s possible to pipe failed tests. So, it wouldn't be a problem to return a non-zero code. 

However, is it conventional to treat a successful output as an error? I think that it’s all about naming. For example, the mentioned `--pm-check-unused-files` option implies 0 (no unused files found) or 1 (found unused files). The proposed `--pm-reveal-unused-files` option kind of allows files not to be found.

>So, probably it should be another option for use in GH actions (or `pre-commit`) `--pm-check-unused-files` that will exit w/ a non-zero exit code to fail a build if any files have been found. 

Another option would be to add the environment variable for CI runs:  `PYTEST_MATCHER_RETURN_CODES=yes pytest tests --pm-reveal-unused-files`.

What do you prefer more?

zaufi commented:
> `PYTEST_MATCHER_RETURN_CODES=yes pytest tests --pm-reveal-unused-files`

This variant LGTM.

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +383,19 @@ def pytest_addoption(parser) -> None:
+@pytest.hookimpl(trylast=True)
+def pytest_configure(config: pytest.Config):
+    """Register additional configuration."""
+    if os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() == 'yes':  # NOQA: SIM108
+        plugin_return_codes = True
+    else:
+        plugin_return_codes = False
```
Why not simple like this?
```suggestion
    plugin_return_codes = os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() == 'yes'
```

zaufi replied:
Oe even like this
```suggestion
    plugin_return_codes = os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() in ('yes', 'true', '1')
```

zaufi replied:
Except for this question, the rest is LGTM.

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +383,16 @@ def pytest_addoption(parser) -> None:
+@pytest.hookimpl(trylast=True)
+def pytest_configure(config: pytest.Config):
+    """Register additional configuration."""
+    return_codes = os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() in ('yes', 'true', '1')
+
+    if config.getoption('--pm-reveal-unused-files'):
```
No need to check the environment unless CLI option has given
```suggestion
    if config.getoption('--pm-reveal-unused-files'):
        return_codes = os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() in ('yes', 'true', '1')
```

zaufi commented on src/matcher/plugin.py:
```diff
@@ -336,4 +383,15 @@ def pytest_addoption(parser) -> None:
+@pytest.hookimpl(trylast=True)
+def pytest_configure(config: pytest.Config):
+    """Register additional configuration."""
+    if config.getoption('--pm-reveal-unused-files'):
+        return_codes = os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() in ('yes', 'true', '1')
+        config.option.collectonly = True
+        reporter = _UnusedFilesReporter(return_codes=return_codes)
+        config.pluginmanager.unregister(name='terminalreporter')
+        config.pluginmanager.register(reporter, 'terminalreporter')
```
Fail fast
```suggestion
    if not config.getoption('--pm-reveal-unused-files'):
        return

    return_codes = os.getenv('PYTEST_MATCHER_RETURN_CODES', '').lower() in ('yes', 'true', '1')
    config.option.collectonly = True
    reporter = _UnusedFilesReporter(return_codes=return_codes)
    config.pluginmanager.unregister(name='terminalreporter')
    config.pluginmanager.register(reporter, 'terminalreporter')
```

zaufi commented:
@xymaxim 
Thank you very much for the great job you've done!
